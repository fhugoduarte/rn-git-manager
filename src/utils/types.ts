export enum gitType {
  bitBucket = 'bitbucket',
  gitHub = 'github',
}
