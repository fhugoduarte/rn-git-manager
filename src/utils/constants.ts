export enum gitHubAPI {
  baseUrl = 'https://api.github.com',
  searchUser = '/search/users',
}

export enum bitBucketAPI {}
