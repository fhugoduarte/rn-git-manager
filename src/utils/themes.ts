const colors = {
  gitColor: '#f1502f',
  gitHub: '#333',
  bitBucket: '#205081',
  black: '#333',
  lightGray: '#f5f5f5',
  white: '#fafafa',
  purple: '#6e5494',
  orange: '#c9510c',
  red: '#bd2c00',
  lightGreen: '#6cc644',
  darkGreen: '#2ebc4f',
  blue: '#4078c0',
};

export default colors;
