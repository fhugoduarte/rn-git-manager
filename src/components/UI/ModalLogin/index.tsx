import React, { useEffect, useRef, useState } from 'react';
import { Animated, TouchableWithoutFeedback, Keyboard } from 'react-native';
import FWIcon from 'react-native-vector-icons/FontAwesome';

import themes from '../../../utils/themes';
import { gitType } from '../../../utils/types';
import {
  Container,
  OverlayContainer,
  ModalContainer,
  ModalContent,
  HeaderContainer,
  Label,
  Input,
  SignInButtonContainer,
  SignInButton,
  SignInButtonTitle,
} from './styles';

const visiblePrevios = (value: boolean): boolean => {
  const ref = useRef(false);
  useEffect(
    (): void => {
      ref.current = value;
    },
  );
  return ref.current;
};

interface Props {
  visible: boolean;
  git: gitType | undefined;
  onClose(): void;
  onSubmit(username: string, password: string, git?: gitType): void;
}

const modalLogin: React.FC<Props> = ({
  visible,
  git,
  onClose,
  onSubmit,
}: Props) => {
  let animatedValue = new Animated.Value(0);
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const [keyboardOpened, setKeyboardStatus] = useState(false);

  const modalRef = useRef<any>(null);
  const prevVisible = visiblePrevios(visible);

  const isBitBucket = git === gitType.bitBucket;
  let keyboardDidShowListener: any = null;
  let keyboardDidHideListener: any = null;

  useEffect((): void => {
    keyboardDidShowListener = Keyboard.addListener(
      'keyboardDidShow',
      (): void => {
        setKeyboardStatus(true);
      },
    );

    keyboardDidHideListener = Keyboard.addListener(
      'keyboardDidHide',
      (): void => {
        setKeyboardStatus(false);
      },
    );
  }, []);

  useEffect(
    (): (() => void) => (): void => {
      keyboardDidShowListener.remove();
      keyboardDidHideListener.remove();
    },
    [],
  );

  const showModal = (): void => {
    animatedValue.setValue(0);
    if (modalRef && modalRef.current) {
      modalRef.current.setNativeProps({ style: { zIndex: 5 } });
    }
    setTimeout((): void => {
      Animated.timing(animatedValue, {
        toValue: 100,
        duration: 500,
        useNativeDriver: true,
      }).start();
    }, 100);
  };

  const hideModal = (): void => {
    animatedValue.setValue(100);
    Animated.timing(animatedValue, {
      toValue: 200,
      duration: 500,
      useNativeDriver: true,
    }).start(
      (): void => {
        if (modalRef && modalRef.current) {
          modalRef.current.setNativeProps({ style: { zIndex: -1 } });
        }
        Keyboard.dismiss();
        setUsername('');
        setPassword('');
      },
    );
  };

  useEffect((): void => {
    const hasOpened = !prevVisible && visible;
    const hasClosed = prevVisible && !visible;

    if (hasOpened) {
      showModal();
    } else if (hasClosed) {
      hideModal();
    }
  }, [visible]);

  const closeModal = (): void => {
    if (keyboardOpened) return Keyboard.dismiss();
    return onClose();
  };

  const submit = (): void => {
    Keyboard.dismiss();
    if (username !== '' && password !== '') onSubmit(username, password, git);
  };

  return (
    <Container
      ref={modalRef}
      style={{
        opacity: animatedValue.interpolate({
          inputRange: [0, 100, 200],
          outputRange: [0, 1, 0],
        }),
      }}
    >
      <TouchableWithoutFeedback onPress={closeModal}>
        <OverlayContainer />
      </TouchableWithoutFeedback>
      <ModalContainer behavior="padding">
        <ModalContent
          style={{
            transform: [
              {
                translateY: animatedValue.interpolate({
                  inputRange: [0, 100, 200],
                  outputRange: [100, 0, -100],
                }),
              },
            ],
          }}
        >
          <HeaderContainer>
            {isBitBucket ? (
              <FWIcon name="bitbucket" size={50} color={themes.bitBucket} />
            ) : (
              <FWIcon name="github" size={50} color={themes.gitHub} />
            )}
          </HeaderContainer>
          <Label>Usu√°rio ou Email</Label>
          <Input value={username} onChangeText={setUsername} />
          <Label>Senha</Label>
          <Input value={password} onChangeText={setPassword} />
          <SignInButtonContainer>
            <SignInButton onPress={submit}>
              <SignInButtonTitle>Entrar</SignInButtonTitle>
            </SignInButton>
          </SignInButtonContainer>
        </ModalContent>
      </ModalContainer>
    </Container>
  );
};

export default modalLogin;
