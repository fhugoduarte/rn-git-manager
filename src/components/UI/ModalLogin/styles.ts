import { Animated, Dimensions } from 'react-native';
import styled from 'styled-components/native';

import themes from '../../../utils/themes';
const { width, height } = Dimensions.get('screen');

const Container = styled(Animated.View)`
  position: absolute;
  align-items: center;
  justify-content: center;
  width: ${width};
  height: ${height};
`;

const OverlayContainer = styled(Animated.View)`
  position: absolute;
  background-color: #00000040;
  width: ${width};
  height: ${height};
`;

const ModalContainer = styled.KeyboardAvoidingView`
  flex: 1;
  position: absolute;
  align-items: center;
  justify-content: center;
`;

const ModalContent = styled(Animated.View)`
  width: ${width * 0.8};
  justify-content: center;
  background-color: white;
  border-radius: 6px;
  border-width: 2px;
  border-color: gray;
  padding: 20px;
`;

const HeaderContainer = styled.View`
  padding: 10px;
  align-items: center;
`;

const Label = styled.Text`
  font-size: 18px;
  font-weight: bold;
  padding: 15px 5px;
`;

const Input = styled.TextInput`
  border-radius: 3px;
  border-width: 1px;
  border-color: gray;
  font-size: 18px;
  padding: 10px;
`;

const SignInButtonContainer = styled.View`
  align-items: center;
  justify-content: center;
`;

const SignInButton = styled.TouchableOpacity`
  align-items: center;
  justify-content: center;
  border-radius: 4px;
  margin-top: 20px;
  padding: 10px 40px;
  align-items: center;
  flex-direction: row;
  background-color: ${themes.darkGreen};
`;

const SignInButtonTitle = styled.Text`
  color: #fff;
  font-size: 18px;
  font-weight: 400;
  padding-left: 15px;
`;

export {
  Container,
  OverlayContainer,
  ModalContainer,
  ModalContent,
  HeaderContainer,
  Label,
  Input,
  SignInButtonContainer,
  SignInButton,
  SignInButtonTitle,
};
