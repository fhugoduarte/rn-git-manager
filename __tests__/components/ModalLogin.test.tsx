import React from 'react';
import renderer from 'react-test-renderer';

import ModalLogin from '../../src/components/UI/ModalLogin';
import { gitType } from '../../src/utils/types';

jest.mock('NativeAnimatedHelper');
jest.useFakeTimers();

describe('Render ModalLogin', (): void => {
  it('should renders correctly with gitHub props', (): void => {
    const tree = renderer
      .create(
        <ModalLogin
          visible
          git={gitType.gitHub}
          onClose={(): void => {}}
          onSubmit={(): void => {}}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should renders correctly with bitBucket props', (): void => {
    const tree = renderer
      .create(
        <ModalLogin
          visible
          git={gitType.bitBucket}
          onClose={(): void => {}}
          onSubmit={(): void => {}}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
