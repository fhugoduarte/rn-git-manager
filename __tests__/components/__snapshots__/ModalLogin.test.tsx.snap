// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render ModalLogin should renders correctly with bitBucket props 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "height": 1334,
      "justifyContent": "center",
      "opacity": 0,
      "position": "absolute",
      "width": 750,
    }
  }
>
  <View
    accessible={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#00000040",
        "height": 1334,
        "position": "absolute",
        "width": 750,
      }
    }
  />
  <View
    onLayout={[Function]}
    style={
      Array [
        Array [
          Object {
            "alignItems": "center",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "position": "absolute",
          },
        ],
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "white",
          "borderColor": "gray",
          "borderRadius": 6,
          "borderWidth": 2,
          "justifyContent": "center",
          "paddingBottom": 20,
          "paddingLeft": 20,
          "paddingRight": 20,
          "paddingTop": 20,
          "transform": Array [
            Object {
              "translateY": 100,
            },
          ],
          "width": 600,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "paddingBottom": 10,
              "paddingLeft": 10,
              "paddingRight": 10,
              "paddingTop": 10,
            },
          ]
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "#205081",
                "fontSize": 50,
              },
              undefined,
              Object {
                "fontFamily": "FontAwesome",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
      </View>
      <Text
        style={
          Array [
            Object {
              "fontSize": 18,
              "fontWeight": "bold",
              "paddingBottom": 15,
              "paddingLeft": 5,
              "paddingRight": 5,
              "paddingTop": 15,
            },
          ]
        }
      >
        Usuário ou Email
      </Text>
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "borderColor": "gray",
              "borderRadius": 3,
              "borderWidth": 1,
              "fontSize": 18,
              "paddingBottom": 10,
              "paddingLeft": 10,
              "paddingRight": 10,
              "paddingTop": 10,
            },
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
      <Text
        style={
          Array [
            Object {
              "fontSize": 18,
              "fontWeight": "bold",
              "paddingBottom": 15,
              "paddingLeft": 5,
              "paddingRight": 5,
              "paddingTop": 15,
            },
          ]
        }
      >
        Senha
      </Text>
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "borderColor": "gray",
              "borderRadius": 3,
              "borderWidth": 1,
              "fontSize": 18,
              "paddingBottom": 10,
              "paddingLeft": 10,
              "paddingRight": 10,
              "paddingTop": 10,
            },
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
            },
          ]
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#2ebc4f",
              "borderRadius": 4,
              "flexDirection": "row",
              "justifyContent": "center",
              "marginTop": 20,
              "opacity": 1,
              "paddingBottom": 10,
              "paddingLeft": 40,
              "paddingRight": 40,
              "paddingTop": 10,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#fff",
                  "fontSize": 18,
                  "fontWeight": "400",
                  "paddingLeft": 15,
                },
              ]
            }
          >
            Entrar
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`Render ModalLogin should renders correctly with gitHub props 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "height": 1334,
      "justifyContent": "center",
      "opacity": 0,
      "position": "absolute",
      "width": 750,
    }
  }
>
  <View
    accessible={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#00000040",
        "height": 1334,
        "position": "absolute",
        "width": 750,
      }
    }
  />
  <View
    onLayout={[Function]}
    style={
      Array [
        Array [
          Object {
            "alignItems": "center",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "position": "absolute",
          },
        ],
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "white",
          "borderColor": "gray",
          "borderRadius": 6,
          "borderWidth": 2,
          "justifyContent": "center",
          "paddingBottom": 20,
          "paddingLeft": 20,
          "paddingRight": 20,
          "paddingTop": 20,
          "transform": Array [
            Object {
              "translateY": 100,
            },
          ],
          "width": 600,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "paddingBottom": 10,
              "paddingLeft": 10,
              "paddingRight": 10,
              "paddingTop": 10,
            },
          ]
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "#333",
                "fontSize": 50,
              },
              undefined,
              Object {
                "fontFamily": "FontAwesome",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
      </View>
      <Text
        style={
          Array [
            Object {
              "fontSize": 18,
              "fontWeight": "bold",
              "paddingBottom": 15,
              "paddingLeft": 5,
              "paddingRight": 5,
              "paddingTop": 15,
            },
          ]
        }
      >
        Usuário ou Email
      </Text>
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "borderColor": "gray",
              "borderRadius": 3,
              "borderWidth": 1,
              "fontSize": 18,
              "paddingBottom": 10,
              "paddingLeft": 10,
              "paddingRight": 10,
              "paddingTop": 10,
            },
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
      <Text
        style={
          Array [
            Object {
              "fontSize": 18,
              "fontWeight": "bold",
              "paddingBottom": 15,
              "paddingLeft": 5,
              "paddingRight": 5,
              "paddingTop": 15,
            },
          ]
        }
      >
        Senha
      </Text>
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "borderColor": "gray",
              "borderRadius": 3,
              "borderWidth": 1,
              "fontSize": 18,
              "paddingBottom": 10,
              "paddingLeft": 10,
              "paddingRight": 10,
              "paddingTop": 10,
            },
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
            },
          ]
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#2ebc4f",
              "borderRadius": 4,
              "flexDirection": "row",
              "justifyContent": "center",
              "marginTop": 20,
              "opacity": 1,
              "paddingBottom": 10,
              "paddingLeft": 40,
              "paddingRight": 40,
              "paddingTop": 10,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#fff",
                  "fontSize": 18,
                  "fontWeight": "400",
                  "paddingLeft": 15,
                },
              ]
            }
          >
            Entrar
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
